mcpserver/tools.go

```diff
@@ -461,68 +484,81 @@ Failure to do so will result in the tool being stuck, awaiting for the command t
 			mcp.Items(map[string]any{"type": "number"}),
 		),
 	),
-	Handler: func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
-		repo, env, err := openEnvironment(ctx, request)
+	ParseRequest: func(request mcp.CallToolRequest) (*EnvironmentRunCmdToolRequest, error) {
+		base, err := parseBaseEnvironmentRequest(request)
 		if err != nil {
-			return mcp.NewToolResultErrorFromErr("unable to open the environment", err), nil
+			return nil, err
 		}
 
-		command := request.GetString("command", "")
-		shell := request.GetString("shell", "sh")
+		req := &EnvironmentRunCmdToolRequest{
+			BaseEnvironmentRequest: base,
+			Command:                request.GetString("command", ""),
+			Shell:                  request.GetString("shell", "sh"),
+			Background:             request.GetBool("background", false),
+			UseEntrypoint:          request.GetBool("use_entrypoint", false),
+		}
 
-		updateRepo := func() (*mcp.CallToolResult, error) {
-			if err := repo.Update(ctx, env, request.GetString("explanation", "")); err != nil {
-				return mcp.NewToolResultErrorFromErr("failed to update repository", err), err
+		if portList, ok := request.GetArguments()["ports"].([]any); ok {
+			for _, port := range portList {
+				req.Ports = append(req.Ports, int(port.(float64)))
```

Comment:
🐛 **Bug (Medium):** Unsafe type assertion to float64 without checking - could cause runtime panic if port is not a float64

```suggestion
if portFloat, ok := port.(float64); ok {
	req.Ports = append(req.Ports, int(portFloat))
} else {
	return nil, fmt.Errorf("invalid port type: %T", port)
}
```